"Heading is descriptive" (b49b2e): replace "section of content" (#1425)
act-rules/act-rules.github.io: _rules/heading-descriptive-b49b2e.md

## Expectation

Each target element describes the topic or purpose of the first [palpable content][] which is:

- after the target, in tree order in the [flat tree][]; and
- [visible][], if the target is [visible][]; and

> - [included in the accessibility tree][], if the target is [included in the accessibility tree][].

Author: This strongly suggest that this rule should be split in two, one for each case…

Reviewer: Not if those 'and' conditional are intentional. I read this as:

If the target is visible and included in the accessibility tree, then it should describe the first palpable content that is visible and included in the accessibility tree that comes after the target in the flat tree.

However, what about non-visible headings describing visible content, where both are included in the accessibility tree? This is not an unusual use case, eg. a news front page with a hidden heading 'Top headlines'. The heading isn't needed visually as that is usually apparent from other cues.

Author: If it is both visible and included in the accessibility tree, then it is "visible or included in the accessibility tree" (the or is inclusive by default). And then it matches both the expectations and need to describe both visible and accessible content (which may be the same or different palpable content).

## Expectation

> Each target element describes the topic or purpose of the first non-[decorative][] [palpable content][] which is:

Reviewer: Good reference find on "palpable content"!!

> Each target element describes the topic or purpose of the first non-[decorative][] [palpable content][] which is:
> 
> - after the target, in tree order in the [flat tree][]; and
> - [visible][], if the target is [visible][]; and
> - [included in the accessibility tree][], if the target is [included in the accessibility tree][].

Reviewer 2: This list isn't following the conditional list format. The reason it can't is because its topic changes. I think you can reformulate this in multiple expectations. I prefer doing this over splitting the rule:

```suggestion
Each target element describes the topic or purpose of the first [non-decorative][decorative] [palpable content][] after the target in tree order in the [flat tree][].

## Expectation 2

Each [palpable content][] from expectation 1 is [visible][], except if its target element is not [visible][].

## Expectation 3

Each [palpable content][] from expectation 1 is [included in the accessibility tree][], except if its target element is not [included in the accessibility tree][].
```

Reviewer: Not sure I understand that.

**Note:** Headings do not need to be lengthy. A word, or even a single character, may suffice.

## Assumptions

- This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content), and sufficiently understood.

> - This rule assumes that the [flat tree][] order is close to the reading order, as elements are rendered on the page. Due to positioning, it is possible to render a document in a order that greatly differ from the tree order, in which case the content which is visually associated with a heading might not be the content following it in tree order and this rule might fail while [Success Criterion 2.4.6 Headings and Label][sc246] is still satisfied.

Reviewer 3: Isn't the rule essentially assuming conformance to 1.3.2 (https://www.w3.org/TR/WCAG/#meaningful-sequence)? In that case, it seems like an even more reasonable assumption to make.

Author: Not really, because the programmatically determined order could be different from the DOM order (and match the visual order). For example by tweaking the accessibility tree with `aria-owns`.

Reviewer 3: Makes sense 👍 I also just looked into CSS Grid and, interestingly enough, grid layouts that re-arrange the logical order of content are non-conforming: https://drafts.csswg.org/css-grid/#order-accessibility

Reviewer: Not to mention the fun that can ensue when flexbox gets involved!

</html>
```

#### Passed Example 8

> This heading describes the first [palpable content][] after it. The next one (the second `p` element) is not considered by this rule.

Reviewer 4: ```suggestion
This heading describes the first [palpable content][] after it (the first `p` element). The next [palpable content][](the second `p` element) is not considered by this rule.
```

</html>
```

#### Failed Example 5

> This heading does not describe the first [palpable content][] after it. The next one (the second `p` element) is not considered by this rule.

Reviewer 4: ```suggestion
This heading does not describe the first [palpable content][] after it- the first `p` element. The next [palpable content][], the second `p` element, is not considered by this rule.
```
